.config/doom/config.org
#+TITLE: BK's doom config
#+STARTUP: content

* Foundation
** evil
#+begin_src elisp
(setq evil-escape-key-sequence "jk")
#+end_src

** Dired
#+begin_src elisp
(setq find-file-visit-truename nil)
#+end_src

* Look & Feel
** Theme
Have been using doom-opera, and it looks good.

#+begin_src elisp
(setq doom-theme 'doom-opera)
;; (setq doom-theme 'doom-one)
#+end_src

** Line numbers
I don't prefer to have line numbers

#+begin_src elisp
(setq display-line-numbers-type nil)
#+end_src

** Fill-column
Set 80 columns for programming. Hide the line for wriitng.

#+begin_src elisp
(setq-default fill-column 80)
(remove-hook! '(text-mode-hook) #'+fill-column-enable-h)
#+end_src

** Word-wrap
#+begin_src elisp
(when (modulep! :editor word-wrap)
  (add-hook! (c-mode-common
              python-mode
              emacs-lisp-mode) #'+word-wrap-mode))
#+end_src

** Fonts
#+begin_src elisp
(setq doom-font (font-spec :family "JetBrains Mono" :size 10.0)
      doom-big-font (font-spec :family "JetBrains Mono" :size 20.0)
      doom-variable-pitch-font (font-spec :family "Noto Sans CJK KR" :size 10.0)
      ;; doom-unicode-font (font-spec :name "NanumGothicCoding" :size 13.0)
      )
(set-fontset-font "fontset-default" 'hangul (font-spec :family "NanumGothicCoding" :size 10.0))
#+end_src

** Hangul input
#+begin_src elisp
(setq default-input-method "korean-hangul")
(map! "S-SPC" #'toggle-input-method)
#+end_src

** RBG
#+begin_src elisp
(when (modulep! :tools rgb)
  (add-hook! 'rainbow-mode-hook
    (hl-line-mode (if rainbow-mode -1 +1))))
#+end_src

** Page break lines
#+begin_src elisp :tangle packages.el
(package! page-break-lines)
#+end_src

#+begin_src elisp
(use-package! page-break-lines
  :hook ((text-mode
          emacs-lisp-mode
          lisp-mode
          compilation-mode
          outline-mode
          help-mode) . page-break-lines-mode))
#+end_src

* Project
** Projectile
#+begin_src elisp
(setq projectile-project-search-path '(("~/Codes/" . 2)))
(setq +workspaces-switch-project-function (lambda (dir) (dired dir)))
(after! projectile
  (projectile-register-project-type 'cmake
                                    '("CMakeLists.txt")
                                    :compilation-dir "build"
                                    :configure "cmake %s"
                                    :compile "cmake --build . -- -j8"
                                    :test "ctest")
  (projectile-register-project-type 'waf
                                    '("wscript")
                                    :configure "waf distclean configure"
                                    :compile "waf")

  (define-key!
    [remap projectile-compile-project] nil))
#+end_src

* Writing
** Org
#+begin_src elisp
(setq org-directory "~/org")
(after! org
  (setq org-tags-column 0))
#+end_src

** Deft
#+begin_src elisp
;; (when (modulep! :ui deft)
;;   (after! deft
;;     (setq deft-directory org-directory)))
#+end_src

* Programming
** LSP
#+begin_src elisp
(when (modulep! :tools lsp)
  (setq lsp-file-watch-threshold 10000)
  (setq lsp-lens-enable nil))

(after! lsp-mode
  (setq lsp-enable-suggest-server-download nil
        lsp-auto-install-server nil))
#+end_src

** CMake (C++)
#+begin_src elisp
;; Doom: config.el
;; (after! eglot
;;   (setq eglot-server-programs
;;         (cl-remove-if
;;          (lambda (it)
;;            (equal (car it) '(cmake-mode cmake-ts-mode)))
;;          eglot-server-programs))
;;   (add-to-list 'eglot-server-programs
;;                '((cmake-mode cmake-ts-mode) . ("neocmakelsp" "--stdio"))))
#+end_src

** CCLS (C++)
#+begin_src elisp
;; (after! ccls
;;   (setq ccls-initialization-options
;;         (append ccls-initialization-options
;;                 `(:compilationDatabaseDirectory "build"))))
#+end_src

** Python
#+begin_src elisp
;; (setq lsp-pyright-venv-directory ".venv"
;;       lsp-pyright-langserver-command "basedpyright")
;; (after! lsp-pyright
;;   (setq lsp-pyright-venv-directory ".venv"
;;         lsp-pyright-langserver-command '("basedpyright-langserver")))
#+end_src

* External Apps & Services
** Org Jira
#+begin_src elisp :tangle packages.el
;; (package! org-jira)
#+end_src

* LLM
** gptel
#+begin_src elisp :tangle packages.el
;; (unpin! gptel)
;; (package! gptel :recipe (:host github :repo "karthink/gptel"))
#+end_src

#+begin_src elisp
;; (use-package! gptel
;;   :config
;;   (setq gptel-backend
;;         (gptel-make-gemini "Gemini" :key #'gptel-api-key :stream t)
;;         gptel-model 'gemini-2.5-flash-preview-05-20))  ;; or 'gemini-2.5-pro
#+end_src


.config/doom/config.el
;;; .doom.d/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here

;; ui fine-tuning
(setq doom-theme 'doom-opera)

;; (setq doom-font (font-spec :family "JetBrains Mono" :size 14)
;;       doom-variable-pitch-font (font-spec :family "Noto Sans CJK KR" :size 14)
;;       doom-unicode-font (font-spec :family "JetBrains Mono")
;;       doom-big-font (font-spec :family "JetBrains Mono" :size 24))
;; (set-fontset-font t 'hangul (font-spec :name "Noto Sans Mono CJK KR"))
(setq doom-font (font-spec :family "JetBrains Mono" :size 14)
      doom-big-font (font-spec :family "JetBrains Mono" :size 24)
      doom-variable-pitch-font (font-spec :family "Noto Sans CJK KR" :size 14)
      doom-unicode-font (font-spec :name "Noto Sans Mono CJK KR"))

;; Korean input methods
(setq default-input-method "korean-hangul")
(map! "S-SPC" #'toggle-input-method)


;; I don't like line numbers
(setq display-line-numbers-type nil)

;; org mode
(setq org-directory "~/Dropbox/Org")
(after! org
  (setq org-tags-column -100))

;; deft
(when (featurep! :ui deft)
  (after! deft
    (setq deft-directory org-directory)))

;; fill column mode not in org mode
(remove-hook! '(text-mode-hook) #'+fill-column-enable-h)

;; fill-column = 80
(setq-default fill-column 80)

;; lsp config not to see warning every time
(when (featurep! :tools lsp)
  (setq lsp-file-watch-threshold 10000))

;; word-wrap config
(when (featurep! :editor word-wrap)
  (add-hook! (c-mode-common
              python-mode
              emacs-lisp-mode) #'+word-wrap-mode))

;; rgb mode config
(when (featurep! :tools rgb)
  (add-hook! 'rainbow-mode-hook
    (hl-line-mode (if rainbow-mode -1 +1))))

;; additional packages
(use-package! page-break-lines
  :hook ((text-mode
          emacs-lisp-mode
          lisp-mode
          compilation-mode
          outline-mode
          help-mode) . page-break-lines-mode))

;; google-c-style
(use-package! google-c-style
  :hook ((c-mode-common . google-set-c-style)
         (c-mode-common . google-make-newline-indent)))

;; projectile configs
(setq projectile-project-search-path '("~/dev/prj" "~/dev/sandbox" "~/dev/ref"))
(after! projectile
  (projectile-register-project-type 'cmake
                                    '("CMakeLists.txt")
                                    :compilation-dir "build"
                                    :configure "cmake %s"
                                    :compile "cmake --build . -- -j8"
                                    :test "ctest")
  (projectile-register-project-type 'waf
                                    '("wscript")
                                    :configure "waf distclean configure"
                                    :compile "waf")

  (define-key!
    [remap projectile-compile-project] nil))

;; ccls set out-of-tree build directory
(after! ccls
  (setq ccls-initialization-options
        (append ccls-initialization-options
                `(:compilationDatabaseDirectory "build"))))

;; iedit only for Emacs key-bindings
(use-package! iedit
  :unless (featurep! :editor evil)
  :bind (("C-;" . iedit-mode)
         ("C-:" . iedit-mode-toggle-on-function))
  :config (set-face-inverse-video 'iedit-occurrence t))

;; emacs additional key-bindings
(unless (featurep! :editor evil)
  (map!
   ;; dired-jump binding is not working before 'dired' is first used
   "C-x k"   #'doom/kill-this-buffer-in-all-windows
   "C-x C-j" #'dired-jump

   ;; projectile-comfile-command should be used instead of ivy/project-compile
   [remap projectile-compile-project]  nil

   (:when (featurep! :completion ivy)
    "C-s" #'counsel-grep-or-swiper
    "C-r" #'counsel-grep-or-swiper-backward)

   (:when (featurep! :tools lsp)
    "C-c C-f"  #'+default/lsp-format-region-or-buffer) ;; this is my coding habit

   (:when (featurep! :editor multiple-cursors)
    "M-3" #'mc/mark-previous-like-this
    "M-4" #'mc/mark-next-like-this
    "M-#" #'mc/skip-to-previous-like-this
    "M-$" #'mc/skip-to-next-like-this)
   )
  )
